sdk/identity/Azure.Identity/README.md

```diff
@@ -109,92 +86,167 @@ using Azure.Security.KeyVault.Secrets;
+|-|-
+|`AZURE_CLIENT_ID`|id of an Azure Active Directory application
+|`AZURE_TENANT_ID`|id of the application's Azure Active Directory tenant
+|`AZURE_CLIENT_SECRET`|one of the application's client secrets
+
+#### Service principal with certificate
+|variable name|value
+|-|-
+|`AZURE_CLIENT_ID`|id of an Azure Active Directory application
+|`AZURE_TENANT_ID`|id of the application's Azure Active Directory tenant
+|`AZURE_CLIENT_CERTIFICATE_PATH`|path to a PEM-encoded certificate file including private key (without password protection)
+
+#### Username and password
+|variable name|value
+|-|-
+|`AZURE_CLIENT_ID`|id of an Azure Active Directory application
+|`AZURE_USERNAME`|a username (usually an email address)
+|`AZURE_PASSWORD`|that user's password
+
+Configuration is attempted in the above order. For example, if values for a
+client secret and certificate are both present, the client secret will be used.
+
+## Troubleshooting
+
+### Error Handling
+Errors arising from authentication can be raised on any service client method which makes a request to the service. This is because the first time the token is requested from the credential is on the first call to the service, and any subsequent calls might need to refresh the token. In order to distinguish these failures from failures in the service client Azure Identity classes raise the `AuthenticationFailedException` with details to the source of the error in the exception message as well as possibly the error message. Depending on the application these errors may or may not be recoverable.
+~~~ c#
 
-This example demonstrates authenticating the `SecretClient` from the [Azure.Security.KeyVault.Secrets][secrets_client_library] client library using the `DeviceCodeCredential`. The sample constructs a `DeviceCodeCredential` with an application client id, and a callback which prints the device code.
-```c#
 using Azure.Identity;
 using Azure.Security.KeyVault.Secrets;
-using System.Threading.Tasks;
 
-Func<DeviceCodeInfo, Task> PrintDeviceCode = code => 
-{ 
-    Console.WriteLine(code.Message);
+// Create a secret client using the DefaultAzureCredential
+var client = new SecretClient(new Uri("https://myvault.azure.vaults.net/"), new DefaultAzureCredential());
 
-    return Task.CompletedTask;
+try
+{
+    KeyVaultSecret secret = await client.GetSecretAsync("secret1");
 }
+catch (AuthenticationFailedException e)
+{
+    Console.WriteLine($"Authentication Failed. {e.Message}");
+    return null;
+}
+~~~
+For more details on dealing with errors arising from failed requests to Azure Active Directory, or managed identity endpoints please refer to the Azure Active Directory [documentation on authorization error codes][aad_err_doc].
 
-// Create a secret client using the DeviceCodeCredential
-var client = new SecretClient(new Uri("https://myvault.azure.vaults.net/"), new DeviceCodeCredential((code, cancellationToken) => PrintDeviceCode(code), clientId));
-```
-## Troubleshooting
+### Logging
```

Comment:
Should we provide sample logs output for DAC cred?
and display how it notified the user of which credential was picked up ?

Reply:
This is a good suggestion. I would like to add this information somewhere, but possibly in a more in depth troubleshooting walkthrough rather than directly in the readme.

Reply:
okay, also how about specifying the difference between error types CredentialUnavailableException and ClientAuthenticationException ?